# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Необходимо выполнить несколько заданий:
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [1]:
import pandas as pd #импортируем библиотеки
from sqlalchemy import create_engine 

In [2]:
#подключаемсяк базе данных
db_config = {'user': 'praktikum_student', # имя пользователя            
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
books = '''
SELECT *
FROM books'''
books = pd.io.sql.read_sql(books, con = engine) #работаем с одной из  таблиц

In [4]:
def showtime(df): #функция для ознакомления с данными
    display("Выведем первые строки  Данных")
    display(df.head())
    display('-'*10)
    display("Выведем информацию по  имеющимся в файле Данным")
    display(df.info())
    display('-'*10)
    display("Есть ли пропуски в  Данных и их количество")
    display(df.isna().sum())
    display('-'*10)
    display("Какое количество дубликатов в  Данных")
    display(df.duplicated().sum())
    


In [5]:
showtime(books) #рассмотрим данные таблицы

'Выведем первые строки  Данных'

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

'----------'

'Выведем информацию по  имеющимся в файле Данным'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

'----------'

'Есть ли пропуски в  Данных и их количество'

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

'----------'

'Какое количество дубликатов в  Данных'

0

1 тыс. строк. Все данные в целочисленном и строковых видах, пропусков нет как и дубликатов

In [6]:
authors = '''
SELECT *
FROM authors'''
authors = pd.io.sql.read_sql(authors, con = engine) #работаем с фреймом авторы

In [7]:
showtime(authors)

'Выведем первые строки  Данных'

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

'----------'

'Выведем информацию по  имеющимся в файле Данным'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

'----------'

'Есть ли пропуски в  Данных и их количество'

author_id    0
author       0
dtype: int64

'----------'

'Какое количество дубликатов в  Данных'

0

636  ненулевых данных, 2 колонки. одна  в целочисленном формате, другая  в виде строки. пропусков нет, дублей тоже.

In [8]:
publishers= '''
SELECT *
FROM publishers'''
publishers = pd.io.sql.read_sql(publishers, con = engine) #работаем с фреймом publishers

In [9]:
showtime(publishers)

'Выведем первые строки  Данных'

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

'----------'

'Выведем информацию по  имеющимся в файле Данным'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

'----------'

'Есть ли пропуски в  Данных и их количество'

publisher_id    0
publisher       0
dtype: int64

'----------'

'Какое количество дубликатов в  Данных'

0

340 строк данных, 2 колонки,  одна в целочисленном формате, другая в виде строки. пропусков нет, дублей тоже.

In [10]:
ratings= '''
SELECT *
FROM ratings'''
ratings = pd.io.sql.read_sql(ratings, con = engine) #работаем с фреймом ratings

In [11]:
showtime(ratings)

'Выведем первые строки  Данных'

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

'----------'

'Выведем информацию по  имеющимся в файле Данным'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

'----------'

'Есть ли пропуски в  Данных и их количество'

rating_id    0
book_id      0
username     0
rating       0
dtype: int64

'----------'

'Какое количество дубликатов в  Данных'

0

6456  строк данных, 4 колонки, данные в целочисленном и одна колонка в виде строки. пропусков нет, дублей тоже.

In [12]:
reviews= '''
SELECT *
FROM reviews'''
reviews = pd.io.sql.read_sql(reviews, con = engine) #работаем с фреймом reviews

In [13]:
showtime(reviews)

'Выведем первые строки  Данных'

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

'----------'

'Выведем информацию по  имеющимся в файле Данным'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

'----------'

'Есть ли пропуски в  Данных и их количество'

review_id    0
book_id      0
username     0
text         0
dtype: int64

'----------'

'Какое количество дубликатов в  Данных'

0

2793 строк данных, 4 колонки, данные в целочисленном  формати и в  виде строки. пропусков нет, дублей тоже.

Выводы: по  всем  таблицам  представлены данные в виде чисел и строк, пропусков нет.

## Выполним задания

 Задача  1: Посчитайте, сколько книг вышло после 1 января 2000 года;

In [14]:
cb =  """

       SELECT COUNT(*) FROM books WHERE publication_date > '2000-01-01';
"""



cb  =  pd.io.sql.read_sql(cb, con = engine)

In [15]:
print(f' После 2020-01-01  года вышло {cb["count"][0]} книг')

 После 2020-01-01  года вышло 819 книг


Задача 2: Для каждой книги посчитайте количество обзоров и среднюю оценку;


In [16]:
m_r = '''
            WITH first AS
                (SELECT books.book_id AS book_id, 
                COUNT(reviews.text) AS counts,
                books.title
                FROM reviews
                RIGHT JOIN books on reviews.book_id = books.book_id
                GROUP BY books.book_id),

                second AS
                (SELECT books.book_id AS book_id, 
                ROUND(AVG(ratings.rating),2) AS mid_rating,
                books.title
                FROM ratings
                RIGHT JOIN books on ratings.book_id = books.book_id
                GROUP BY books.book_id) 

                SELECT second.book_id AS book_id,
                second.title AS title,
                first.counts AS counts,
                round(second.mid_rating,2) AS rating                    
                FROM first
                INNER JOIN second ON first.book_id = second.book_id
                ''' 

m_r = pd.io.sql.read_sql(m_r, con = engine)    


In [17]:
print('Количество обзоров и средняя оценка  каждой книги')

Количество обзоров и средняя оценка  каждой книги


In [18]:
print(m_r)

     book_id                                              title  counts  \
0          1                                       'Salem's Lot       2   
1          2                 1 000 Places to See Before You Die       1   
2          3  13 Little Blue Envelopes (Little Blue Envelope...       3   
3          4  1491: New Revelations of the Americas Before C...       2   
4          5                                               1776       4   
..       ...                                                ...     ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)       3   
996      997                        Xenocide (Ender's Saga  #3)       3   
997      998                                    Year of Wonders       4   
998      999                        You Suck (A Love Story  #2)       2   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...       4   

     rating  
0      3.67  
1      2.50  
2      4.67  
3      4.50  
4      4.00  
..      ...  
9

Задача 3.Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;


In [19]:
b_p = '''
          SELECT
          publishers.publisher_id AS id,
          publishers.publisher AS name,
          COUNT(DISTINCT(books.book_id)) AS count

          FROM
          publishers

          INNER JOIN books ON books.publisher_id = publishers.publisher_id

          WHERE
          books.num_pages > 50

          GROUP BY publishers.publisher_id, publishers.publisher
          ORDER BY count DESC

          '''
b_p = pd.io.sql.read_sql(b_p, con = engine)

In [20]:
print(f'Издательство, которое выпустило наибольшее число книг толще 50 страниц  - {b_p.loc[0,"name"]}')

Издательство, которое выпустило наибольшее число книг толще 50 страниц  - Penguin Books


Задача 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [21]:
author = '''
            
         SELECT
         chart.author_id AS author_id,
         chart.author_name AS author_name,
         ROUND(AVG(chart.middle),2) AS rating
            FROM

         (SELECT    
          books.author_id AS author_id,
          authors.author AS author_name,
          ratings.book_id AS book_id,
          ROUND(AVG(ratings.rating),2) AS middle   
             FROM
          ratings
          
          INNER JOIN
          books ON books.book_id = ratings.book_id
          
          INNER JOIN
          authors ON books.author_id = authors.author_id    
          GROUP BY ratings.book_id, books.author_id, authors.author
              HAVING
          COUNT(ratings.rating_id) >= 50) AS chart
          GROUP BY chart.author_id, chart.author_name
          ORDER BY rating DESC

          '''
author = pd.io.sql.read_sql(author, con = engine)

In [22]:
author

author_id                                        author_name  rating
0         236                         J.K. Rowling/Mary GrandPré    4.29
1         240                                     J.R.R. Tolkien    4.26
2         402                  Markus Zusak/Cao Xuân Việt Khương    4.26
3         376                                  Louisa May Alcott    4.19
4         498                                       Rick Riordan    4.08
5         621                                    William Golding    3.90
6         235                                      J.D. Salinger    3.83
7         630  William Shakespeare/Paul Werstine/Barbara A. M...    3.79
8         469           Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.79
9         106                                          Dan Brown    3.76
10        372                                         Lois Lowry    3.75
11        195            George Orwell/Boris Grabnar/Peter Škerl    3.73
12        554                                    Stephenie Meyer    3.66
13        311                                     John Steinbeck    3.62

In [23]:
print(f' автор {author.loc[0,"author_name"]} с самой высокой средней оценкой книг -  {author.loc[0,"rating"]}')

 автор J.K. Rowling/Mary GrandPré с самой высокой средней оценкой книг -  4.29


Задача 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


In [24]:
m_mark = '''
         SELECT 
         ROUND(AVG(review.number)) AS mid_review
           FROM
         (SELECT 
          username AS name,
          COUNT(review_id) AS number
            FROM 
          reviews

            WHERE 
          username IN
          (SELECT 
           username AS name
             FROM 
           ratings

           GROUP BY name
            HAVING 
           COUNT(rating_id) > 50)
           GROUP BY name) AS review

                            '''

m_mark = pd.io.sql.read_sql(m_mark, con=engine)


In [25]:
m_mark

mid_review
0        24.0

In [26]:
print(f'среднее количество обзоров -  {m_mark.loc[0,"mid_review"]} от пользователей, которые поставили больше 50 оценок')

среднее количество обзоров -  24.0 от пользователей, которые поставили больше 50 оценок


1. После 2020-01-01  года вышло 819 книг
2. Для каждой книги посчитали количество обзоров и среднюю оценку;
3. Издательство, которое выпустило наибольшее число книг толще 50 страниц  - Penguin Books
4. Автор J.K. Rowling/Mary GrandPré с самой высокой средней оценкой книг -  4.29
5. Среднее количество обзоров -  24.0 от пользователей, которые поставили больше 50 оценок